crates/c-api/src/ref.rs

```diff
@@ -182,25 +186,100 @@ pub extern "C" fn wasm_foreign_new(_store: &crate::wasm_store_t) -> Box<wasm_for
     abort("wasm_foreign_new")
 }
 
-pub type wasmtime_anyref_t = ManuallyRooted<AnyRef>;
-pub type wasmtime_externref_t = ManuallyRooted<ExternRef>;
+/// C-API representation of `anyref`.
+///
+/// This represented differently in the C API from the header to handle how
+/// this is dispatched internally. Null anyref values are represented with a
+/// `store_id` of zero, and otherwise the `rooted` field is valid.
+///
+/// Note that this relies on the Wasmtime definition of `ManuallyRooted` to have
+/// a 64-bit store_id first.
+pub union wasmtime_anyref_t {
+    store_id: u64,
+    rooted: ManuallyDrop<ManuallyRooted<AnyRef>>,
+}
+
+/// Same as `wasmtime_anyref_t`, but for extenref.
+pub union wasmtime_externref_t {
+    store_id: u64,
+    rooted: ManuallyDrop<ManuallyRooted<ExternRef>>,
+}
```

Comment:
A `ManuallyRooted<T>` already contains a store ID, is there a way we can avoid duplicating it here?

Reply:
Ah nevermind I see now that this is a union, not a struct, and that `store_id == 0` is used as `None`.

Reply:
But, could we just make this `Option<ManuallyDrop<ManuallyRooted<ExternRef>>>` instead of a union?

Reply:
That's a good point about `Option`, but one caveat is that embedders need to be able to make a null reference, test for one, and ideally without calling an API functions. That's why I opted for this representation where with `#[repr(C)]` we put the `store_id` first everywhere and can test it for zero here (and it's `NonZeroU64` in `StoreId`). 

So in that sense we could indeed use `Option` here, but it would rely on having a static assertion along the lines of:


```rust
    union T {
        a: u64,
        b: std::mem::ManuallyDrop<Option<ManuallyRooted<ExternRef>>>,
    }
    unsafe {
        assert!(
            T {
                b: std::mem::ManuallyDrop::new(None)
            }
            .a == 0
        );
    }
```

which, in testing, does in fact work. I'm sort of 50/50 though on whether this is worth it or "too clever by half"... What do you think, worth it to rely on that niche optimization?

Reply:
I feel like this is worth doing, and keeping the union inside just the assertion instead of permeating the rest of the Rust code. I don't think this is too niche of an optimization because I was under the impression that it was "well known" that Rust would pack discriminants into non-zero and non-null things.

Reply:
The other option here is to have a

```rust
impl ManuallyRooted<T> {
    #[doc(hidden)]
    pub fn into_raw_parts_for_c_api(&self) -> (u64, u32) { ... }
}
```

method and a dual `from_raw_parts_for_c_api` method.

Reply:
That might be simpler and easier to maintain and ensure remains correct long term.

Reply:
Good point about the `#[doc(hidden)]` bits, I like that route too so I've implemented that